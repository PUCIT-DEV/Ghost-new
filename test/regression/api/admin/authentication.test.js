const {expect} = require('chai');
const {agentProvider, mockManager, fixtureManager, matchers} = require('../../../utils/e2e-framework');
const {anyEtag, anyDate, anyErrorId} = matchers;

const security = require('@tryghost/security');
const models = require('../../../../core/server/models');
const settingsCache = require('../../../../core/shared/settings-cache');

// Requires needed to enable a labs flag
const sinon = require('sinon');
const configUtils = require('../../../utils/configUtils');

describe('Authentication API', function () {
    let agent;

    describe('Blog setup', function () {
        before(async function () {
            agent = await agentProvider.getAdminAPIAgent();
        });

        beforeEach(function () {
            mockManager.mockMail();
        });

        afterEach(function () {
            mockManager.restore();
        });

        it('is setup? no', async function () {
            await agent
                .get('authentication/setup')
                .expectStatus(200)
                .matchBodySnapshot()
                .matchHeaderSnapshot({
                    etag: anyEtag
                });
        });

        it('complete setup', async function () {
            // Enable the improvedOnboarding flag
            configUtils.set('enableDeveloperExperiments', true);
            sinon.stub(settingsCache, 'get');
            settingsCache.get.withArgs('labs').returns({
                improvedOnboarding: true
            });
            settingsCache.get.callThrough();

            await agent
                .post('authentication/setup')
                .body({
                    setup: [{
                        name: 'test user',
                        email: 'test@example.com',
                        password: 'thisissupersafe',
                        blogTitle: 'a test blog',
                        theme: 'TryGhost/Dawn'
                    }]
                })
                .expectStatus(201)
                .matchBodySnapshot({
                    users: [{
                        created_at: anyDate,
                        updated_at: anyDate
                    }]
                })
                .matchHeaderSnapshot({
                    etag: anyEtag
                });

            // Test our side effects
            mockManager.assert.sentEmail({
                subject: 'Your New Ghost Site',
                to: 'test@example.com'
            });

            expect(await settingsCache.get('active_theme')).to.eq('dawn');
        });

        it('is setup? yes', async function () {
            await agent
                .get('authentication/setup')
                .matchBodySnapshot()
                .matchHeaderSnapshot({
                    etag: anyEtag
                });
        });

        it('complete setup again', function () {
            return agent
                .post('authentication/setup')
                .body({
                    setup: [{
                        name: 'test user',
                        email: 'test-leo@example.com',
                        password: 'thisissupersafe',
                        blogTitle: 'a test blog'
                    }]
                })
                .expectStatus(403)
                .matchBodySnapshot({
                    errors: [{
                        id: anyErrorId
                    }]
                })
                .matchHeaderSnapshot({
                    etag: anyEtag
                });
        });

        it('update setup', async function () {
            await fixtureManager.init();
            await agent.loginAsOwner();

            await agent
                .put('authentication/setup')
                .body({
                    setup: [{
                        name: 'test user edit',
                        email: 'test-edit@example.com',
                        password: 'thisissupersafe',
                        blogTitle: 'a test blog'
                    }]
                })
                .expectStatus(200)
                .matchBodySnapshot({
                    users: [{
                        created_at: anyDate,
                        last_seen: anyDate,
                        updated_at: anyDate
                    }]
                })
                .matchHeaderSnapshot({
                    etag: anyEtag
                });
        });
    });

    describe('Invitation', function () {
        before(async function () {
            agent = await agentProvider.getAdminAPIAgent();
            await fixtureManager.init('invites');
            await agent.loginAsOwner();
        });

        it('check invite with invalid email', function () {
            return agent
                .get('authentication/invitation?email=invalidemail')
                .expectStatus(400)
                .matchHeaderSnapshot({
                    etag: anyEtag
                });
        });

        it('check valid invite', async function () {
            await agent
                .get(`authentication/invitation?email=${fixtureManager.get('invites', 0).email}`)
                .expectStatus(200)
                .matchBodySnapshot()
                .matchHeaderSnapshot({
                    etag: anyEtag
                });
        });

        it('check invalid invite', async function () {
            await agent
                .get(`authentication/invitation?email=notinvited@example.org`)
                .expectStatus(200)
                .matchBodySnapshot()
                .matchHeaderSnapshot({
                    etag: anyEtag
                });
        });

        it('try to accept without invite', function () {
            return agent
                .post('authentication/invitation')
                .body({
                    invitation: [{
                        token: 'lul11111',
                        password: 'lel123456',
                        email: 'not-invited@example.org',
                        name: 'not invited'
                    }]
                })
                .expectStatus(404)
                .matchHeaderSnapshot({
                    etag: anyEtag
                });
        });

        it('try to accept with invite and existing email address', function () {
            return agent
                .post('authentication/invitation')
                .body({
                    invitation: [{
                        token: fixtureManager.get('invites', 0).token,
                        password: '12345678910',
                        email: fixtureManager.get('users', 0).email,
                        name: 'invited'
                    }]
                })
                .expectStatus(422)
                .matchBodySnapshot({
                    errors: [{
                        id: anyErrorId
                    }]
                })
                .matchHeaderSnapshot({
                    etag: anyEtag
                });
        });

        it('try to accept with invite', async function () {
            await agent
                .post('authentication/invitation')
                .body({
                    invitation: [{
                        token: fixtureManager.get('invites', 0).token,
                        password: '12345678910',
                        email: fixtureManager.get('invites', 0).email,
                        name: 'invited'
                    }]
                })
                .expectStatus(200)
                .matchBodySnapshot()
                .matchHeaderSnapshot({
                    etag: anyEtag
                });
        });
    });

    describe('Password reset', function () {
        const email = fixtureManager.get('users', 0).email;

        before(async function () {
            agent = await agentProvider.getAdminAPIAgent();
            // NOTE: this order of fixture initialization boggles me. Ideally should not depend on agent/login sequence
            await fixtureManager.init('invites');
            await agent.loginAsOwner();
        });

        beforeEach(function () {
            mockManager.mockMail();
        });

        afterEach(function () {
            mockManager.restore();
        });

        it('reset password', async function () {
            const ownerUser = await fixtureManager.getCurrentOwnerUser();

            const token = security.tokens.resetToken.generateHash({
                expires: Date.now() + (1000 * 60),
                email: email,
                dbHash: settingsCache.get('db_hash'),
                password: ownerUser.get('password')
            });

            const res = await agent.put('authentication/passwordreset')
                .header('Accept', 'application/json')
                .body({
                    passwordreset: [{
                        token: token,
                        newPassword: 'thisissupersafe',
                        ne2Password: 'thisissupersafe'
                    }]
                })
                .expectStatus(200)
                .matchBodySnapshot()
                .matchHeaderSnapshot({
                    etag: anyEtag
                });
        });

        it('reset password: invalid token', async function () {
            await agent
                .put('authentication/passwordreset')
                .header('Accept', 'application/json')
                .body({
                    passwordreset: [{
                        token: 'invalid',
                        newPassword: 'thisissupersafe',
                        ne2Password: 'thisissupersafe'
                    }]
                })
                .expectStatus(401)
                .matchBodySnapshot({
                    errors: [{
                        id: anyErrorId
                    }]
                })
                .matchHeaderSnapshot({
                    etag: anyEtag
                });
        });

        it('reset password: expired token', async function () {
            const ownerUser = await fixtureManager.getCurrentOwnerUser();

            const dateInThePast = Date.now() - (1000 * 60);
            const token = security.tokens.resetToken.generateHash({
                expires: dateInThePast,
                email: email,
                dbHash: settingsCache.get('db_hash'),
                password: ownerUser.get('password')
            });

            await agent
                .put('authentication/passwordreset')
                .header('Accept', 'application/json')
                .body({
                    passwordreset: [{
                        token: token,
                        newPassword: 'thisissupersafe',
                        ne2Password: 'thisissupersafe'
                    }]
                })
                .expectStatus(400)
                .matchBodySnapshot({
                    errors: [{
                        id: anyErrorId
                    }]
                })
                .matchHeaderSnapshot({
                    etag: anyEtag
                });
        });

        it('reset password: unmatched token', async function () {
            const token = security.tokens.resetToken.generateHash({
                expires: Date.now() + (1000 * 60),
                email: email,
                dbHash: settingsCache.get('db_hash'),
                password: 'invalid_password'
            });

            await agent
                .put('authentication/passwordreset')
                .header('Accept', 'application/json')
                .body({
                    passwordreset: [{
                        token: token,
                        newPassword: 'thisissupersafe',
                        ne2Password: 'thisissupersafe'
                    }]
                })
                .expectStatus(400)
                .matchBodySnapshot({
                    errors: [{
                        id: anyErrorId
                    }]
                })
                .matchHeaderSnapshot({
                    etag: anyEtag
                });
        });

        it('reset password: generate reset token', async function () {
            await agent
                .post('authentication/passwordreset')
                .header('Accept', 'application/json')
                .body({
                    passwordreset: [{
                        email: email
                    }]
                })
                .expectStatus(200)
                .matchBodySnapshot()
                .matchHeaderSnapshot({
                    etag: anyEtag
                });
        });
    });

    describe('Reset all passwords', function () {
        before(async function () {
            agent = await agentProvider.getAdminAPIAgent();
            await fixtureManager.init('invites');
            await agent.loginAsOwner();
        });

        beforeEach(function () {
            mockManager.mockMail();
        });

        afterEach(function () {
            mockManager.restore();
        });

        it('reset all passwords returns 200', async function () {
            await agent.post('authentication/reset_all_passwords')
                .header('Accept', 'application/json')
                .body({})
                .expectStatus(200)
                .matchBodySnapshot()
                .matchHeaderSnapshot({
                    etag: anyEtag
                });

            // Check side effects
            // All users locked
            const users = await models.User.fetchAll();
            for (const user of users) {
                expect(user.get('status')).to.equal('locked');
            }

            // No session left
            const sessions = await models.Session.fetchAll();
            expect(sessions.length).to.equal(0);

            mockManager.assert.sentEmailCount(2);

            mockManager.assert.sentEmail({
                subject: 'Reset Password',
                to: 'jbloggs@example.com'
            });
            mockManager.assert.sentEmail({
                subject: 'Reset Password',
                to: 'ghost-author@example.com'
            });
        });
    });
});
