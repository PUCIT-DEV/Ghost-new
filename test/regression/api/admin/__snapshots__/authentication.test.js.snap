// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authentication API Blog setup complete setup 1: [body] 1`] = `
Object {
  "users": Array [
    Object {
      "accessibility": null,
      "bio": null,
      "cover_image": null,
      "created_at": Any<String>,
      "email": "test@example.com",
      "facebook": null,
      "id": "1",
      "last_seen": null,
      "location": null,
      "meta_description": null,
      "meta_title": null,
      "name": "test user",
      "profile_image": null,
      "slug": "test",
      "status": "active",
      "tour": null,
      "twitter": null,
      "updated_at": Any<String>,
      "url": "http://127.0.0.1:2369/author/joe-bloggs/",
      "website": null,
    },
  ],
}
`;

exports[`Authentication API Blog setup complete setup 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "434",
  "content-type": "application/json; charset=utf-8",
  "etag": Any<String>,
  "vary": "Origin, Accept-Encoding",
  "x-cache-invalidate": "/*",
  "x-powered-by": "Express",
}
`;

exports[`Authentication API Blog setup complete setup again 1: [body] 1`] = `
Object {
  "errors": Array [
    Object {
      "code": null,
      "context": null,
      "details": null,
      "help": null,
      "id": Any<String>,
      "message": "Setup has already been completed.",
      "property": null,
      "type": "NoPermissionError",
    },
  ],
}
`;

exports[`Authentication API Blog setup complete setup again 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "201",
  "content-type": "application/json; charset=utf-8",
  "etag": Any<String>,
  "vary": "Origin, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Authentication API Blog setup is setup? no 1: [body] 1`] = `
Object {
  "setup": Array [
    Object {
      "status": false,
    },
  ],
}
`;

exports[`Authentication API Blog setup is setup? no 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "28",
  "content-type": "application/json; charset=utf-8",
  "etag": Any<String>,
  "vary": "Origin, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Authentication API Blog setup is setup? yes 1: [body] 1`] = `
Object {
  "setup": Array [
    Object {
      "status": true,
    },
  ],
}
`;

exports[`Authentication API Blog setup is setup? yes 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "27",
  "content-type": "application/json; charset=utf-8",
  "etag": Any<String>,
  "vary": "Origin, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Authentication API Blog setup update setup 1: [body] 1`] = `
Object {
  "users": Array [
    Object {
      "accessibility": null,
      "bio": "bio",
      "cover_image": null,
      "created_at": Any<String>,
      "email": "test-edit@example.com",
      "facebook": null,
      "id": "1",
      "last_seen": Any<String>,
      "location": "location",
      "meta_description": null,
      "meta_title": null,
      "name": "test user edit",
      "profile_image": "https://example.com/super_photo.jpg",
      "slug": "test",
      "status": "active",
      "tour": null,
      "twitter": null,
      "updated_at": Any<String>,
      "url": "http://127.0.0.1:2369/author/joe-bloggs/",
      "website": null,
    },
  ],
}
`;

exports[`Authentication API Blog setup update setup 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "506",
  "content-type": "application/json; charset=utf-8",
  "etag": Any<String>,
  "vary": "Origin, Accept-Encoding",
  "x-cache-invalidate": "/*",
  "x-powered-by": "Express",
}
`;

exports[`Authentication API Invitation check invalid invite 1: [body] 1`] = `
Object {
  "invitation": Array [
    Object {
      "valid": false,
    },
  ],
}
`;

exports[`Authentication API Invitation check invalid invite 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "32",
  "content-type": "application/json; charset=utf-8",
  "etag": Any<String>,
  "vary": "Origin, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Authentication API Invitation check invite with invalid email 1: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "206",
  "content-type": "application/json; charset=utf-8",
  "etag": Any<String>,
  "vary": "Origin, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Authentication API Invitation check valid invite 1: [body] 1`] = `
Object {
  "invitation": Array [
    Object {
      "valid": true,
    },
  ],
}
`;

exports[`Authentication API Invitation check valid invite 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "31",
  "content-type": "application/json; charset=utf-8",
  "etag": Any<String>,
  "vary": "Origin, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Authentication API Invitation try to accept with invite 1: [body] 1`] = `
Object {
  "invitation": Array [
    Object {
      "message": "Invitation accepted.",
    },
  ],
}
`;

exports[`Authentication API Invitation try to accept with invite 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "51",
  "content-type": "application/json; charset=utf-8",
  "etag": Any<String>,
  "vary": "Origin, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Authentication API Invitation try to accept with invite and existing email address 1: [body] 1`] = `
Object {
  "errors": Array [
    Object {
      "code": null,
      "context": "Attempting to create an account with existing email address.",
      "details": null,
      "help": "Use different email address to register your account.",
      "id": Any<String>,
      "message": "Could not create an account, email is already in use.",
      "property": null,
      "type": "ValidationError",
    },
  ],
}
`;

exports[`Authentication API Invitation try to accept with invite and existing email address 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "328",
  "content-type": "application/json; charset=utf-8",
  "etag": Any<String>,
  "vary": "Origin, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Authentication API Invitation try to accept without invite 1: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "181",
  "content-type": "application/json; charset=utf-8",
  "etag": Any<String>,
  "vary": "Origin, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Authentication API Password reset reset password 1: [body] 1`] = `
Object {
  "passwordreset": Array [
    Object {
      "message": "Password changed successfully.",
    },
  ],
}
`;

exports[`Authentication API Password reset reset password 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "64",
  "content-type": "application/json; charset=utf-8",
  "etag": Any<String>,
  "vary": "Origin, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Authentication API Password reset reset password: expired token 1: [body] 1`] = `
Object {
  "errors": Array [
    Object {
      "code": null,
      "context": "Password reset link expired.",
      "details": null,
      "help": "Request a new password reset via the login form.",
      "id": Any<String>,
      "message": "Cannot reset password.",
      "property": null,
      "type": "BadRequestError",
    },
  ],
}
`;

exports[`Authentication API Password reset reset password: expired token 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "260",
  "content-type": "application/json; charset=utf-8",
  "etag": Any<String>,
  "vary": "Origin, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Authentication API Password reset reset password: generate reset token 1: [body] 1`] = `
Object {
  "passwordreset": Array [
    Object {
      "message": "Check your email for further instructions.",
    },
  ],
}
`;

exports[`Authentication API Password reset reset password: generate reset token 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "76",
  "content-type": "application/json; charset=utf-8",
  "etag": Any<String>,
  "vary": "Origin, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Authentication API Password reset reset password: invalid token 1: [body] 1`] = `
Object {
  "errors": Array [
    Object {
      "code": null,
      "context": "Invalid password reset link.",
      "details": null,
      "help": "Check if password reset link has been fully copied or request new password reset via the login form.",
      "id": Any<String>,
      "message": "Cannot reset password.",
      "property": null,
      "type": "UnauthorizedError",
    },
  ],
}
`;

exports[`Authentication API Password reset reset password: invalid token 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "314",
  "content-type": "application/json; charset=utf-8",
  "etag": Any<String>,
  "vary": "Origin, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Authentication API Password reset reset password: unmatched token 1: [body] 1`] = `
Object {
  "errors": Array [
    Object {
      "code": null,
      "context": "Password reset link has already been used.",
      "details": null,
      "help": "Request a new password reset via the login form.",
      "id": Any<String>,
      "message": "Cannot reset password.",
      "property": null,
      "type": "BadRequestError",
    },
  ],
}
`;

exports[`Authentication API Password reset reset password: unmatched token 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "274",
  "content-type": "application/json; charset=utf-8",
  "etag": Any<String>,
  "vary": "Origin, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Authentication API Reset all passwords reset all passwords returns 200 1: [body] 1`] = `Object {}`;

exports[`Authentication API Reset all passwords reset all passwords returns 200 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "2",
  "content-type": "application/json; charset=utf-8",
  "etag": Any<String>,
  "vary": "Origin, Accept-Encoding",
  "x-powered-by": "Express",
}
`;
